src/main/java/net/fabricmc/loader/impl/discovery/ModSolver.java

```diff
@@ -0,0 +1,980 @@
+				false, installableMods, true,
+				dependencyHelper);
+
+		if (!dependencyHelper.hasASolution()) {
+			Log.warn(LogCategory.RESOLUTION, "Unable to find a solution to fix the mod set, reason: %s", dependencyHelper.why());
+			return null;
+		}
+
+		List<ModCandidate> activeMods = new ArrayList<>();
+		List<AddModVar> modsToAdd = new ArrayList<>();
+		List<ModCandidate> modsToRemove = new ArrayList<>();
+		Map<AddModVar, List<ModCandidate>> modReplacements = new HashMap<>();
+
+		for (DomainObject obj : dependencyHelper.getASolution()) {
+			if (obj instanceof ModCandidate) {
+				activeMods.add((ModCandidate) obj);
+			} else if (obj instanceof AddModVar) {
+				List<ModCandidate> replaced = new ArrayList<>();
+
+				ModCandidate selectedMod = selectedMods.get(obj.getId());
+				if (selectedMod != null) replaced.add(selectedMod);
+
+				List<ModCandidate> mods = modsById.get(obj.getId());
+
+				if (mods != null) replaced.addAll(mods);
+
+				if (replaced.isEmpty()) {
+					modsToAdd.add((AddModVar) obj);
+				} else {
+					modReplacements.put((AddModVar) obj, replaced);
+				}
+			} else if (obj instanceof RemoveModVar) {
+				boolean found = false;
+
+				ModCandidate mod = selectedMods.get(obj.getId());
+
+				if (mod != null) {
+					modsToRemove.add(mod);
+					found = true;
+				}
+
+				List<ModCandidate> mods = modsById.get(obj.getId());
+
+				if (mods != null) {
+					for (ModCandidate m : mods) {
+						if (m.isRoot()) {
+							modsToRemove.add(m);
+							found = true;
+						}
+					}
+				}
+
+				assert found;
+			} else { // unexpected domainobj kind
+				assert false : obj;
+			}
+		}
+
+		// TODO: test if the solution is actually valid?
+
+		return new Fix(modsToAdd, modsToRemove, modReplacements, activeMods);
+	}
+
+	static long fixSolveTime;
+
+	private static Version deriveVersion(VersionBounds vb) {
+		if (!vb.isSemantic()) {
+			return vb.getMin() != null ? vb.getMin() : vb.getMax();
+		}
+
+		SemanticVersion v = (SemanticVersion) vb.getMin();
+
+		if (v != null) { // min bound present
+			if (!vb.isMinInclusive()) { // not inclusive, increment slightly
+				String pr = v.getPrereleaseKey().orElse(null);
+				int[] comps = ((SemanticVersionImpl) v).getVersionComponents();
+
+				if (pr != null) { // has prerelease, add to increase
+					pr = pr.isEmpty() ? "0" : pr.concat(".0");
+				} else { // regular version only, increment patch and make least prerelease
+					if (comps.length < 3) {
+						comps = Arrays.copyOf(comps, comps.length + 1);
+					}
+
+					comps[comps.length - 1]++;
+					pr = "";
+				}
+
+				v = new SemanticVersionImpl(comps, pr, null);
+			}
+		} else if ((v = (SemanticVersion) vb.getMax()) != null) { // max bound only
+			if (!vb.isMaxInclusive()) { // not inclusive, decrement slightly
+				String pr = v.getPrereleaseKey().orElse(null);
+				int[] comps = ((SemanticVersionImpl) v).getVersionComponents();
+
+				if (pr == null) { // no prerelease, use large pr segment
+					pr = "zzzzzzzz";
```

Comment:
If said mod is present, the condition a line above fails and selects another code path, otherwise the least mod version suggested to be installed will be slightly off.